Sources/NIOCore/AsyncSequences/NIOAsyncWriter.swift

```diff
@@ -0,0 +1,750 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2022 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import Atomics
+import NIOConcurrencyHelpers
+
+#if compiler(>=5.5.2) && canImport(_Concurrency)
+/// The delegate of the ``NIOAsyncWriter``.
+@available(macOS 10.15, iOS 13, tvOS 13, watchOS 6, *)
+public protocol NIOAsyncWriterDelegate: Sendable {
+    /// The `Element` type of the delegate and the writer.
+    associatedtype Element: Sendable
+    /// The `Failure` type of the delegate and the writer.
+    associatedtype Failure: Error = Never
+
+    /// This method is called once a sequence was yielded to the ``NIOAsyncWriter``.
+    ///
+    /// If the ``NIOAsyncWriter`` was writable when the sequence was yielded, the sequence will be forwarded
+    /// right away to the delegate. If the ``NIOAsyncWriter`` was _NOT_ writable then the sequence will be buffered
+    /// until the ``NIOAsyncWriter`` becomes writable again.
+    func didYield<S: Sequence>(contentsOf sequence: S) where S.Element == Element
+
+    /// This method is called once the ``NIOAsyncWriter`` is terminated.
+    ///
+    /// Termination happens if:
```

Comment:
Is this always called exactly once?

Reply:
Yes this is guaranteed to be called exactly one same as the `didTerminate` on the `NIOAsyncSeuqenceProducerDelegate`

Reply:
Can we document that?